File: problems/binary-search/0034-find-first-and-last-position-in-sorted-array.js
/**
 * #34 Find First and Last Position of Element in Sorted Array
 * Pattern: lower bound and upper bound binary searches.
 * Time: O(log n) | Space: O(1)
 *
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
function searchRange(nums, target) {
  function bound(strict) {
    let left = 0
    let right = nums.length

    while (left < right) {
      const middle = left + Math.floor((right - left) / 2)
      const shouldMoveRight = strict
        ? nums[middle] <= target
        : nums[middle] < target
      if (shouldMoveRight) left = middle + 1
      else right = middle
    }

    return left
  }

  const first = bound(false)
  if (first === nums.length || nums[first] !== target) return [-1, -1]
  return [first, bound(true) - 1]
}

module.exports = searchRange

File: problems/binary-search/0034-find-first-and-last-position-in-sorted-array.py
"""#34 Find First and Last Position of Element in Sorted Array
Use lower_bound and upper_bound to find the half-open target range.
Time: O(log n) | Space: O(1)
"""


class Solution:
    def searchRange(self, nums, target):
        def lower_bound(value):
            left, right = 0, len(nums)
            while left < right:
                middle = left + (right - left) // 2
                if nums[middle] < value:
                    left = middle + 1
                else:
                    right = middle
            return left

        first = lower_bound(target)
        if first == len(nums) or nums[first] != target:
            return [-1, -1]

        after_last = lower_bound(target + 1)
        return [first, after_last - 1]
